feat(big-number): allow fallback to last available value and fix time range for trend lines

ktmud opened the pull request:
🏆 Enhancements

The Big Number chart has two problems related to the handling of null values:

1. **Misleading time range**: When there is no data at the beginning or ending periods of the filtered time range, the backend returns no data records for those periods, therefore the trend line will only start rendering from the datetime where there is data. This often leads to confusion or misinterpretations, especially then in the dashboards that have multiple trend line charts filtered to the same time range. They could all be showing similar smooth lines, yet representing very different time ranges.
2. **NULL value at the last record**: When the last record returns NULL value for the metric, the Big Number text will render as "No data", but the trend line still renders. Some users have requested to fallback displaying the Big Number with the last available dates.

This PR adds option `timeRangeFixed` to fix the first problem and adds a "fallback to last available value" behavior to address the second problem.

The `timeRangedFixed` option is disabled by default. When enabled, the empty starting and ending periods will take space on the x-axis, but there will be no connected lines, dots, or tooltips, just as if the metric values are NULL.

In case the last record has a NULL value for the metric, we will go back to the last available value and display it in grayed out style (see screen shots below).

Superset backend needs to be updated to return `query.from_dttm` and `query.to_dttm`. Frontend needs to be updated to add two checkbox controls for these options ( https://github.com/apache/incubator-superset/pull/9341). 

This PR also rewrote the Big Number chart plugin with Typescript.

### Screenshot

#### Before

![Snip20200316_62](https://user-images.githubusercontent.com/335541/76821723-fd111a80-67cb-11ea-9f18-b16b232b22fc.png)

#### After

![Snip20200316_59](https://user-images.githubusercontent.com/335541/76821354-f3d37e00-67ca-11ea-8556-38796b45f2f9.png)

![image](https://user-images.githubusercontent.com/335541/76895498-5e2d0280-684d-11ea-839f-375a7277fb2c.png)

Above chart has a time range between 2001 to 2014, yet only since 2006 it starts to have data. When `timeRangeFixed` is enabled, it will render from the beginning of the parsed filter time range (2001), instead of when it starts to actually have data (2006).


#### In Superset

##### Before 

![image](https://user-images.githubusercontent.com/335541/77504560-ec8d1f80-6e1d-11ea-966b-deefde9706b0.png)

##### After

![Snip20200325_15](https://user-images.githubusercontent.com/335541/77595602-dd0de500-6eb6-11ea-91ae-ecd28dc17b9c.png)

I also increased the default font sizes a little bit.


##### After

![toggle-fixed-time-range](https://user-images.githubusercontent.com/335541/77221585-256c8200-6b08-11ea-8bdf-cd8255d8ed37.gif)

### Test Plan

Go to storybook and make sure all examples for Big Number chart work.

There is also a new unit test to ensure `timeRangeUseFallback` works.

### Related

https://github.com/apache-superset/superset-ui-plugins/pull/402
https://github.com/ktmud/superset-ui-plugins/pull/2

## Reviewer
cc @kristw @etr2460

ktmud committed:
feat(big-number): add option to align time range

In Superset, when a timeseries query has no data at the beginning period
or end period of the filtered time range, there will not no data records
at those periods, hence the trendline in Big Number chart would not
render those periods. This often causes confusion and misinterpretaiton
in dashboards, especially for those with multiple trendline charts
aligned with each other. They could all be a very smooth line, but
actually showing very different time ranges.

This PR adds an option "alignTimeRange" to apply the filtered time
range on the xAxis. Date periods for empty data will be rendered, but
there will be no connected lines, dots, or tooltips for them.

It's possible to still show tooltips for those periods, but I decided
not to do that as: 1) it makes things much more complicated; 2) I don't
want to confuse zero or nulls with empty data.

ktmud committed:
fix(big-number): disable alignRange by default

ktmud committed:
refactor(big-number): migrate to Typescript

ktmud committed:
fix(big-number): typescript build

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/BigNumber.tsx:
```diff
@@ -185,21 +191,33 @@ class BigNumberVis extends React.PureComponent {
+      useFixedTimeRange,
     } = this.props;
 
+    // Apply a fixed tange range if a time range is specified.
```
typo `tange` => `time`

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/package.json:
```diff
@@ -42,5 +41,8 @@
     "react": "^15 || ^16"
+  },
+  "devDependencies": {
+    "@types/shortid": "0.0.29"
```
Let's put this in `dependencies`
https://stackoverflow.com/questions/45176661/how-do-i-decide-whether-types-goes-into-dependencies-or-devdependencies

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/BigNumber.tsx:
```diff
@@ -41,65 +43,69 @@ const PROPORTION = {
+
+  constructor(props: BigNumberVisProps) {
     super(props);
     this.gradientId = shortid.generate();
```
could do 

```ts
private gradientId: string = shortid.generate();
```

and omit the constructor

kristw reviewed (commented):
There are `eslint` errors.

```
/home/travis/build/apache-superset/superset-ui-plugins/packages/superset-ui-legacy-preset-chart-big-number/src/index.js
  1:49  error  Missing file extension "ts" for "./BigNumber/index"       import/extensions
  2:54  error  Missing file extension "ts" for "./BigNumberTotal/index"  import/extensions
/home/travis/build/apache-superset/superset-ui-plugins/packages/superset-ui-legacy-preset-chart-big-number/src/preset.js
  20:34  error  Missing file extension "ts" for "./BigNumber"       import/extensions
  21:39  error  Missing file extension "ts" for "./BigNumberTotal"  import/extensions
```

Don't add the `ts` extension like it suggests. Just apply `eslint-disable` because once these file are transpiled into `js` in `lib` and `esm`, having the `.ts` extension in `import` statement will make it fail to find the file.

ktmud committed:
fix(big-number): change tooltip trigger; fix storybook

ktmud committed:
fix(big-number): move @types to dependencies

ktmud committed:
fix(big-number): move all files to ts

ktmud commented:
> Don't add the `ts` extension like it suggests. Just apply `eslint-disable` because once these file are transpiled into `js` in `lib` and `esm`, having the `.ts` extension in `import` statement will make it fail to find the file.

I moved all files to `ts` instead and made Storybook read `index.ts` as well: https://github.com/apache-superset/superset-ui-plugins/pull/403/files#diff-55dda99e8a651cb4fdb5eb443cb1f8b5R50

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/package.json:
```diff
@@ -29,11 +29,11 @@
     "shortid": "^2.2.14"
   },
   "peerDependencies": {
+    "d3-color": "^1.2.3",
```
Let's keep this one as `dependencies`.

ktmud replied:
Any reason why would it be better? I thought it's preferable to have it in `peerDependencies` as long as it is in the `dependencies` of `superset-frontend`.

kristw replied:
Usually `peerDependencies` should be use when only one instance of that lib should exist in the entire app. This could be due to the use of singleton or because the dependency library is very heavy. The downside of `peerDependencies` is it put more responsibility for the app developers who are consuming the library.

* Need to add that library to the app-level `dependencies`. 
* If every lib lists everything as `peerDependencies`, the app will have a long list of `dependencies` that many of them are not used directly in the app code.
* If this lib stop using `d3-color`, app developers need to remove it from their app's `dependencies` or it will become redundant dependencies.
* If the app needs `d3-color` version that is not compatible with this library, then there is a conflict.

So if the library is small and does not have singleton requirement, we should list as `dependencies` and rely on `npm` or `yarn` to handle that. This will encapsulate the dependencies and avoid the additional task for app developers to manage `d3-color` as its `dependencies`.

ktmud replied:
Understood. And this totally makes sense. I'm wondering whether we can issue an "official" list of `peerDependencies` supported by Superset that plugin developers can trust will always be available and stable. I'd imagine `d3` packages to be one of them.

kristw replied:
> "official" list of peerDependencies supported by Superset that plugin developers can trust will always be available and stable. I'd imagine d3 packages to be one of them.

Each plugin may be developed on current version of d3 but then later on if d3 upgrades over time, but some plugins did not migrate, some did, it will cause problem what version Superset should say official, because everybody tries to point to the same d3. At this point, it is not clear what value we actually gain from having d3 as `peerDependencies`. It is better to use `dependencies`. 

Also, these plugins can also be used in another app that is not Superset.

ktmud replied:
I'd say packing multi versions of d3 to the client side is still something we should actively try to avoid. Maybe incorporate something like [InspectPack](https://github.com/formidablelabs/inspectpack/) would help.

ktmud committed:
build(big-number): add @types/d3-color as dependency

ktmud committed:
refactor(big-number): remove renderTooltip as prop

ktmud committed:
feat(big-number): add timeRangeUseFallback options and some refactor

ktmud committed:
fix(big-number): update formatting functions

ktmud committed:
fix(big-number): update copy for no data

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/transformProps.ts:
```diff
@@ -0,0 +1,132 @@
+  bigNumber = data.length === 0 ? null : data[0][metricName];
+
+  if (data.length > 0) {
+    const sortedData = [...data]
```
Could use `data.map` directly instead of `[...data].map` because `.map` already create a copy.

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/transformProps.ts:
```diff
@@ -0,0 +1,132 @@
+    mainColor = color.rgb(r, g, b).hex();
+  }
+
+  let bigNumber = null;
```
Can merge with line 61

```ts
let bigNumber = data.length === 0 ? null : data[0][metricName];
```

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/transformProps.ts:
```diff
@@ -0,0 +1,132 @@
+    timeGrainSqla: granularity,
+    vizType,
+    timeRangeFixed = false,
+    timeRangeUseFallback = false,
```
Perhaps rename this field and change from `boolean` to `bigNumberMode: 'last' | 'lastAvailable'`.
We could extend this to support additional modes: `'average'`, `'sum'` in the future.

ktmud replied:
In that case, should we use a `SelectControl` in Superset, too? I'm a little hesitant to add this for now. Allowing `sum`, `average` or any other computation mode significantly increases the complexity of how the product can be used and may cause confusion among end users. Even if the chart builders know what they are doing, the readers might not. Chart builders will need to add a note to the dashboard explaining what the big number is, or we'd have to do that somewhere in the chart.

kristw replied:
I am just afraid to add this `timeRangeUseFallback` field and soon have to write `db:migration` for it. 
At least, could rename it to `useLastAvailableDataPoint` to be more specific.

For more context: 

The current `BigNumber` and `BigNumberTotal` are behaving differently. One show last value and one show sum. 
There was a feature request to clarify this at one point. 

The returned data from the legacy API are also different.

`BigNumber`: array of value over time

```
[
  {
    __timestamp: 1388534400000.0,
    sum__SP_POP_TOTL: 7237260256.0,
  },
  {
    __timestamp: 1356998400000.0,
    sum__SP_POP_TOTL: 7151135481.0,
  },
  {
    __timestamp: 1325376000000.0,
    sum__SP_POP_TOTL: 7066007165.0,
  },
  {
    __timestamp: 1293840000000.0,
    sum__SP_POP_TOTL: 6984252419.0,
  }
]
```

`BigNumberTotal`: sum as single entry in the array

```
[
  {
    sum__num: 32546308,
  },
]
```

ktmud replied:
Hmm. Why does this sound intuitive and surprising at the same time... 😂 Still, I worry it will be probably even more confusing if Big Number with Trend Line can display sums, too.

kristw replied:
Another alternative solution is to fallback to last available data point by default, but add warning to indicate that the last data point is missing.

kristw replied:
options:
1. current solution but rename variable from `timeRangeUseFallback` to `useLastAvailableDataPoint` (or something that is more descriptive).
2. fallback to last available data point by default, but add warning to indicate that the last data point is missing. (maybe your current solution 1. is already better, or can circle back with @etr2460)
3. changes from boolean to string. 

I'm good with any one them.

ktmud replied:
I like your alternative solution. Maybe display the Big Number grayed out and add a tooltip explaining this is the last available value. It saves the trouble for future migration and makes sure chart readers understand what the value is, too.

ktmud replied:
![image](https://user-images.githubusercontent.com/335541/77501534-8f8d6b80-6e15-11ea-9e4b-15aa6a44fee8.png)

Come up with this: fallback to the last available value by default + add some hover tooltip to the fallback values.

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/transformProps.ts:
```diff
@@ -0,0 +1,132 @@
+    subheaderFontSize,
+    timeGrainSqla: granularity,
+    vizType,
+    timeRangeFixed = false,
```
Is this the old `useFixedTimeRange` field?

ktmud replied:
Yes

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/BigNumber.tsx:
```diff
@@ -185,32 +191,65 @@ class BigNumberVis extends React.PureComponent {
+    // XYChart checks the existence of `domain` property decide whether to apply
+    // a domain or not, so it must not be `null` or `undefined`
+    const xScale: { type: string; domain?: number[] } = { type: 'timeUtc' };
+    const tooltipData = trendLineData?.sort(datum => datum.x);
```
* This `.sort()` will mutate the original `props`, which violates `react` 's one-way data flow concept. Should make a copy first `trendlineData?.concat().sort(...)`
* It seems like `trendLineData` should already have been sorted in `transformProps`?

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/BigNumber.tsx:
```diff
@@ -185,32 +191,65 @@ class BigNumberVis extends React.PureComponent {
+    const xScale: { type: string; domain?: number[] } = { type: 'timeUtc' };
+    const tooltipData = trendLineData?.sort(datum => datum.x);
+    if (tooltipData && timeRangeFixed && fromDatetime) {
+      let { toDatetime } = this.props;
```
```ts
const toDateTime = this.props.toDateTime ?? Date.now();
```

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/BigNumber.tsx:
```diff
@@ -185,32 +191,65 @@ class BigNumberVis extends React.PureComponent {
+      xScale.domain = [fromDatetime, toDatetime];
+    }
+    // if can't find any non-null values, no point rendering the trendline
+    if (!tooltipData?.some(d => d.y !== null)) {
```
Can move this `if` block above the block before it.

kristw reviewed (commented):
Agree with the direction of the solution.

ktmud committed:
fix(big-number): address PR feedbacks

ktmud committed:
feat(big-number): replace timeRangeUseFallback with bigNumberFallback

ktmud committed:
fix: upgrade @types/react-bootstrap

ktmud committed:
build(big-number): move react-bootstrap to dependencies

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/BigNumber.tsx:
```diff
@@ -0,0 +1,309 @@
+  private gradientId: string = shortid.generate();
+
+  static defaultProps = {
+    bigNumberFallback: null,
```
Do not need to specify `null` as default props and use default behavior `undefined`. 
`null` will also do not match the type `bigNumberFallback?: TimeSeriesDatum;` which is `TimeSeriesDatum | undefined`.

Same with `fromDatetime`, `toDatetime`, and `trendLineData`

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/BigNumber.tsx:
```diff
@@ -0,0 +1,309 @@
+  bigNumberFallback?: TimeSeriesDatum;
+  formatNumber: NumberFormatter;
+  formatTime: TimeFormatter;
+  fromDatetime: number;
```
Should this be `fromDatetime?` since the field is optional ? 
Same with `toDatetime` and `trendLineData`

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/src/BigNumber/BigNumber.tsx:
```diff
@@ -0,0 +1,309 @@
+
+import './BigNumber.css';
+
+const defaultNumberFormatter = getNumberFormatter(undefined);
```
Is it possible to omit `undefined`?

ktmud replied:
Just wanted to be explicit. `getNumberFormatter` sounds like something that needs a parameter.

ktmud committed:
refactor(big-number): more coherent types

ktmud committed:
feat(big-number): use alert box for fallback values

kristw commented:
Very nice. Could we move the big number up a bit in case of fallback so they align?
That's last thing and I'll merge.
![Pasted_Image_3_27_20__4_50_PM](https://user-images.githubusercontent.com/1659771/77809015-1bd4a400-704b-11ea-9f28-325e15ceb461.png)

ktmud committed:
build(big-number): remove react-bootstrap

ktmud committed:
build: upgrade nimbus and fix versions

Keep running into building errors locally, so upgrade nimbus and
fix all related packages to the working latest version.

ktmud committed:
feat(big-number): adjust fallback warning alignment

kristw commented on package.json:
```diff
@@ -42,12 +42,12 @@
-    "@airbnb/config-prettier": "^2.0.4",
-    "@airbnb/config-typescript": "^2.1.2",
-    "@airbnb/nimbus": "^2.1.3",
+    "@airbnb/config-babel": "2.2.4",
```
Why pin to specific version without `^`?

ktmud replied:
Was having trouble building. Couldn't figure out why. Just don't want to deal with the unpredictability. Sometimes even with the implied "not a breaking change" in a minor version upgrade, things might still break.

kristw replied:
This repo already check in `yarn.lock` so if you did not modify these package version, it should be the same version that was working in `master`. Could you try reseting `yarn.lock` and `package.json` to `master`'s version while keeping the changes you made to `BigNumber`'s `package.json` and `yarn install` again?

ktmud replied:
Weird, it works... I'm pretty sure I tried that yesterday... Maybe it's because I inadvertently pined `@types/shortid`.  

I think it's probably not a bad idea to pin `devDependencies` as they are unlikely to cause dependency hell and the added stability is valuable. Renovate actually [pin the dependencies for you](https://github.com/ktmud/incubator-superset/pull/2) once you set it up.

kristw commented on packages/superset-ui-legacy-preset-chart-big-number/package.json:
```diff
@@ -29,8 +29,9 @@
   "dependencies": {
     "@data-ui/xy-chart": "^0.0.84",
+    "@types/d3-color": "^1.2.2",
+    "@types/shortid": "0.0.29",
```
same here

kristw replied:
This should use the `^`

ktmud committed:
build: use a non-fixed version for @types/shortid

ktmud committed:
build: revert package versions in main
